# Coleta de dados estruturados
- A base de dados refere-se a informações de; vacinas e quantidade de vacinados no mundo

In [1]:
# importando bibliotecas que serão utilizadas
import pandas as pd
import os
from datetime import datetime

# Importando dados do arquivo


In [2]:
DIRETORIO = 'H:\\Meu Drive\\BI\\Python e Outros\\MBA\\coleta_dados_estruturados\\datasets'   
arquivo = 'country_vaccinations.csv'
dados = os.path.join(DIRETORIO, arquivo) # caminho completo do arquivo

# Verifica o caminho do diretório que o arquivo esta salvo

In [3]:
dados

'H:\\Meu Drive\\BI\\Python e Outros\\MBA\\coleta_dados_estruturados\\datasets\\country_vaccinations.csv'

#  Entrar no diretório onde está o arquivo e transformar em um dataframe do pandas

In [4]:
dataframe = pd.read_csv(dados, sep=',', encoding='latin-1')

In [5]:
dataframe.head(5)

country iso_code        date  total_vaccinations  people_vaccinated  \
0  Afghanistan      AFG  2021-02-22                 0.0                0.0   
1  Afghanistan      AFG  2021-02-23                 NaN                NaN   
2  Afghanistan      AFG  2021-02-24                 NaN                NaN   
3  Afghanistan      AFG  2021-02-25                 NaN                NaN   
4  Afghanistan      AFG  2021-02-26                 NaN                NaN   

   people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
0                      NaN                     NaN                 NaN   
1                      NaN                     NaN              1367.0   
2                      NaN                     NaN              1367.0   
3                      NaN                     NaN              1367.0   
4                      NaN                     NaN              1367.0   

   total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
0                             0.0                            0.0   
1                             NaN                            NaN   
2                             NaN                            NaN   
3                             NaN                            NaN   
4                             NaN                            NaN   

   people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
0                                  NaN                             NaN   
1                                  NaN                            34.0   
2                                  NaN                            34.0   
3                                  NaN                            34.0   
4                                  NaN                            34.0   

                                            vaccines  \
0  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
1  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
2  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
3  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
4  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                 source_name            source_website  
0  World Health Organization  https://covid19.who.int/  
1  World Health Organization  https://covid19.who.int/  
2  World Health Organization  https://covid19.who.int/  
3  World Health Organization  https://covid19.who.int/  
4  World Health Organization  https://covid19.who.int/

# Analisando os dados

In [6]:
# dados referente ao Brasil
dataframe.query('country == "Brazil"').sort_values('date', ascending=False)[:5]

country iso_code        date  total_vaccinations  people_vaccinated  \
11085  Brazil      BRA  2022-03-29         413559595.0        181078067.0   
11084  Brazil      BRA  2022-03-28         412177025.0        180881628.0   
11083  Brazil      BRA  2022-03-27         411548202.0        180800024.0   
11082  Brazil      BRA  2022-03-26         411116197.0        180757995.0   
11081  Brazil      BRA  2022-03-25         410860641.0        180726798.0   

       people_fully_vaccinated  daily_vaccinations_raw  daily_vaccinations  \
11085              160272858.0               1382570.0            656803.0   
11084              160008924.0                628823.0            546999.0   
11083              159842858.0                432005.0            587488.0   
11082              159756178.0                255556.0            548819.0   
11081              159673499.0                684805.0            596456.0   

       total_vaccinations_per_hundred  people_vaccinated_per_hundred  \
11085                          193.26                          84.62   
11084                          192.61                          84.53   
11083                          192.32                          84.49   
11082                          192.12                          84.47   
11081                          192.00                          84.45   

       people_fully_vaccinated_per_hundred  daily_vaccinations_per_million  \
11085                                74.90                          3069.0   
11084                                74.77                          2556.0   
11083                                74.70                          2745.0   
11082                                74.65                          2565.0   
11081                                74.62                          2787.0   

                                                vaccines  \
11085  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11084  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11083  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11082  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   
11081  Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...   

                                           source_name  \
11085  State governments via coronavirusbra1.github.io   
11084  State governments via coronavirusbra1.github.io   
11083  State governments via coronavirusbra1.github.io   
11082  State governments via coronavirusbra1.github.io   
11081  State governments via coronavirusbra1.github.io   

                          source_website  
11085  https://coronavirusbra1.github.io  
11084  https://coronavirusbra1.github.io  
11083  https://coronavirusbra1.github.io  
11082  https://coronavirusbra1.github.io  
11081  https://coronavirusbra1.github.io

In [7]:
# numero de paises distintos na base
dataframe['country'].nunique()

223

# Qual o país com maior número de pessoas imunizadas ?

In [8]:
# identificando a data mais recente da coleta
ultima_data = dataframe['date'].max()
ultima_data

'2022-03-29'

In [9]:
# filtrando a última data
ultima_data = dataframe[dataframe.date == ultima_data]

In [10]:
# quantidade de linhas e colunas dentro do dataset
ultima_data.shape

(54, 15)

# Atenção
Como temos 223 paíes e quando buscamos a data mais recente, só aparecem 54, teremos que fazer outro tipo de filtro no dataset, pois, acabaremos excluindo dados de ourtos países.

In [11]:
dataframe.groupby('country')['date'].max().sort_values(ascending=False)

country
Lebanon                            2022-03-29
Lithuania                          2022-03-29
Kosovo                             2022-03-29
Kazakhstan                         2022-03-29
Japan                              2022-03-29
                                      ...    
Pitcairn                           2021-09-07
Turkmenistan                       2021-09-05
Bonaire Sint Eustatius and Saba    2021-09-01
Saint Helena                       2021-05-05
Falkland Islands                   2021-04-14
Name: date, Length: 223, dtype: object

In [12]:
# agrupar o total de vacinação por país de forma decrescente
dataframe.groupby('country')['total_vaccinations'].max().sort_values(ascending=False).map('{:.0f}'.format)

country
China               3263129000
India               1834500657
United States        560181791
Brazil               413559595
Indonesia            377108938
                       ...    
Falkland Islands          4407
Montserrat                4211
Niue                      4161
Tokelau                   1936
Pitcairn                    94
Name: total_vaccinations, Length: 223, dtype: object

# Top 10 paises com maior número de vacinados

In [13]:
imunizados_por_pais = dataframe.groupby('country')['total_vaccinations'].max().sort_values(ascending=False).map('{:.0f}'.format)
imunizados_por_pais

country
China               3263129000
India               1834500657
United States        560181791
Brazil               413559595
Indonesia            377108938
                       ...    
Falkland Islands          4407
Montserrat                4211
Niue                      4161
Tokelau                   1936
Pitcairn                    94
Name: total_vaccinations, Length: 223, dtype: object

In [14]:
# identificando o tipo de estrutura do objeto, no caso semelhante a uma matriz ou coluna em uma planilha
type(imunizados_por_pais)

pandas.core.series.Series

In [15]:
# transformando o objeto acima em um dataframe, redefinindo e criando uma coluna para o indíce
imunizados = imunizados_por_pais.to_frame().reset_index()
imunizados

country total_vaccinations
0               China         3263129000
1               India         1834500657
2       United States          560181791
3              Brazil          413559595
4           Indonesia          377108938
..                ...                ...
218  Falkland Islands               4407
219        Montserrat               4211
220              Niue               4161
221           Tokelau               1936
222          Pitcairn                 94

[223 rows x 2 columns]

In [16]:
imunizados.sort_values('total_vaccinations', ascending=False).head(10)

country total_vaccinations
87                         Jordan            9727982
21                      Argentina           96504666
152  Democratic Republic of Congo             964948
88                      Nicaragua            9611936
222                      Pitcairn                 94
22                          Spain           93807121
153                      Maldives             913264
154                       Lesotho             905438
89                       Zimbabwe            9039729
155                        Guyana             889522

# Qual vacina é mais usada no mundo ?

In [17]:
# selecionando campos para a análise
vacinas_por_pais = dataframe[['country', 'date', 'vaccines']]
vacinas_por_pais

country        date  \
0      Afghanistan  2021-02-22   
1      Afghanistan  2021-02-23   
2      Afghanistan  2021-02-24   
3      Afghanistan  2021-02-25   
4      Afghanistan  2021-02-26   
...            ...         ...   
86507     Zimbabwe  2022-03-25   
86508     Zimbabwe  2022-03-26   
86509     Zimbabwe  2022-03-27   
86510     Zimbabwe  2022-03-28   
86511     Zimbabwe  2022-03-29   

                                                vaccines  
0      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
1      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
2      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
3      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
4      Johnson&Johnson, Oxford/AstraZeneca, Pfizer/Bi...  
...                                                  ...  
86507  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86508  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86509  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86510  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  
86511  Oxford/AstraZeneca, Sinopharm/Beijing, Sinovac...  

[86512 rows x 3 columns]

In [18]:
# vacinas mais usadas nos países
vacinas_por_pais.groupby('vaccines')['country'].nunique().sort_values(ascending=False)

vaccines
Oxford/AstraZeneca                                                                                    20
Johnson&Johnson, Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                                         17
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech                                                          15
Oxford/AstraZeneca, Pfizer/BioNTech                                                                   11
Johnson&Johnson, Moderna, Novavax, Oxford/AstraZeneca, Pfizer/BioNTech                                 8
                                                                                                      ..
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinopharm/Beijing, Sinovac                               1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac, Sputnik Light, Sputnik V, ZF2001                1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac, Sputnik V                                       1
Moderna, Oxford/AstraZeneca, Pfizer/BioNTech, 

In [19]:
# transformando a série acima em um dataframe, resetando e incluíndo a coluna de indíce
vacinas_pais = dataframe.groupby('country')['vaccines'].unique()
vacinas_pais = vacinas_pais.to_frame().reset_index()

# Qual o país possui o maior número de vacinados por dia ?

In [20]:
dataframe.groupby('country')['daily_vaccinations'].max().sort_values(ascending=False)

country
China            22424286.0
India            10037995.0
Bangladesh        3758404.0
United States     3506960.0
Pakistan          2175773.0
                    ...    
Saint Helena          165.0
Niue                   87.0
Montserrat             53.0
Tokelau                23.0
Pitcairn                1.0
Name: daily_vaccinations, Length: 223, dtype: float64

# Qual a média de vacinação por país

In [21]:
media_pais = dataframe.groupby('country')['daily_vaccinations'].mean().sort_values(ascending=False).map('{:.0f}'.format)
media_pais = media_pais.to_frame().reset_index()
media_pais

country daily_vaccinations
0                China            6930368
1                India            4175994
2        United States            1191727
3               Brazil             943529
4            Indonesia             846289
..                 ...                ...
218  Wallis and Futuna                 34
219            Tokelau                 17
220               Niue                 16
221         Montserrat                 11
222           Pitcairn                  1

[223 rows x 2 columns]

# Dataframes criados

1.Imunizados

2.Vacinas_por_pais

3.media_pais

Com base nos dataframes criados vamos criar uma nova base de dados utilizando os dados que tratamos anteriormente

In [22]:
# Renomeando novas colunas das datas
# vamos inicar alterando o nome de algumas colunas

media_pais = media_pais.rename(columns={'country' : 'pais', 'daily_vaccinations' : 'media_vacinacao'})

In [23]:
media_pais

pais media_vacinacao
0                China         6930368
1                India         4175994
2        United States         1191727
3               Brazil          943529
4            Indonesia          846289
..                 ...             ...
218  Wallis and Futuna              34
219            Tokelau              17
220               Niue              16
221         Montserrat              11
222           Pitcairn               1

[223 rows x 2 columns]

In [24]:
vacinas_pais = vacinas_pais.rename(columns={'country' : 'pais', 'vaccines' : 'vacina_utilizada'})
vacinas_pais[:5]

pais                                   vacina_utilizada
0  Afghanistan  [Johnson&Johnson, Oxford/AstraZeneca, Pfizer/B...
1      Albania  [Oxford/AstraZeneca, Pfizer/BioNTech, Sinovac,...
2      Algeria  [Oxford/AstraZeneca, Sinopharm/Beijing, Sinova...
3      Andorra     [Moderna, Oxford/AstraZeneca, Pfizer/BioNTech]
4       Angola                               [Oxford/AstraZeneca]

In [25]:
imunizados = imunizados.rename(columns={'country' : 'pais', 'total_vaccinations' : 'imunizados'})
imunizados[:5]

pais  imunizados
0          China  3263129000
1          India  1834500657
2  United States   560181791
3         Brazil   413559595
4      Indonesia   377108938

# Criando um novo dataset

In [26]:
dados_tratados = pd.merge(imunizados, media_pais, how='inner')

In [27]:
dados_tratados = pd.merge(dados_tratados, vacinas_pais)
dados_tratados

pais  imunizados media_vacinacao  \
0               China  3263129000         6930368   
1               India  1834500657         4175994   
2       United States   560181791         1191727   
3              Brazil   413559595          943529   
4           Indonesia   377108938          846289   
..                ...         ...             ...   
218  Falkland Islands        4407              75   
219        Montserrat        4211              11   
220              Niue        4161              16   
221           Tokelau        1936              17   
222          Pitcairn          94               1   

                                      vacina_utilizada  
0    [CanSino, Sinopharm/Beijing, Sinopharm/Wuhan, ...  
1             [Covaxin, Oxford/AstraZeneca, Sputnik V]  
2          [Johnson&Johnson, Moderna, Pfizer/BioNTech]  
3    [Johnson&Johnson, Oxford/AstraZeneca, Pfizer/B...  
4    [Johnson&Johnson, Moderna, Novavax, Oxford/Ast...  
..                                                 ...  
218                               [Oxford/AstraZeneca]  
219                               [Oxford/AstraZeneca]  
220                                  [Pfizer/BioNTech]  
221                                  [Pfizer/BioNTech]  
222                               [Oxford/AstraZeneca]  

[223 rows x 4 columns]

# Exportando dados tratados

In [28]:
agora = datetime.now()
try:
    dados_tratados.to_csv('H:\\Meu Drive\\BI\\Python e Outros\\MBA\\coleta_dados_estruturados\\datasets\\statistica.csv', index=True, sep=';')
    
    print(f"Arquivo Gerado com Sucesso em {DIRETORIO} em {agora.strftime('%Y-%m-%d %H:%M:%S')}")
    
except Exception as e:
    
    print("Ocorreu um erro ao gerar o Arquivo:", str(e))

Arquivo Gerado com Sucesso em H:\Meu Drive\BI\Python e Outros\MBA\coleta_dados_estruturados\datasets em 2024-04-04 12:23:10
